Account for forwards which whose corresponding backwards are not invoked

eellison opened the pull request:
Stack from [ghstack](https://github.com/ezyang/ghstack) (oldest at bottom):
* #98341
* #98254
* #97440
* __->__ #98112
* #98101
* #98097

Previously, when we would run a forward graph whose backward we never invoked it would prevent us from switching from warmup to recording. Now, refine the heuristic to allow incrementing the generation as soon as we invoke a backward graph. This still handles the 
```
mod1 = torch.compile(...)

mod2 = torch.compile(...) 

mod2(mod1(x)).sum().backward()
```
case while accounting for graphs which we may not run backward of. 

It also now handles the case where we skip cudagraphify the backward of a forward.


cc @soumith @voznesenskym @penguinwu @anijain2305 @EikanWang @jgong5 @Guobing-Chen @XiaobingSuper @zhuhaozhe @blzheng @Xia-Weiwen @wenzhe-nrv @jiayisunx @peterbell10 @desertfire

eellison committed:
Account for forwards which whose corresponding backwards are not invoked

[ghstack-poisoned]

ezyang commented:
Test?

ezyang commented:
i guess you removed review request so imma wait

eellison commented:
@ezyang, yea, adding test

eellison committed:
Update on "Account for forwards which whose corresponding backwards are not invoked"


Previously, we would run a forward graph whose backward we never invoked. This would prevent us from switching from warmup to recording. Instead, refine the heuristic to allow incrementing the generation as soon as we invoke a backward graph. This still handles the 
```
mod1 = torch.compile(...)

mod2 = torch.compile(...) 

mod2(mod1(x)).sum().backward()
```
case while accounting for graphs which we may not run backward of.


cc soumith voznesenskym penguinwu anijain2305 EikanWang jgong5 Guobing-Chen XiaobingSuper zhuhaozhe blzheng Xia-Weiwen wenzhe-nrv jiayisunx peterbell10 desertfire

[ghstack-poisoned]

eellison committed:
Update on "Account for forwards which whose corresponding backwards are not invoked"


Previously, we would run a forward graph whose backward we never invoked. This would prevent us from switching from warmup to recording. Instead, refine the heuristic to allow incrementing the generation as soon as we invoke a backward graph. This still handles the 
```
mod1 = torch.compile(...)

mod2 = torch.compile(...) 

mod2(mod1(x)).sum().backward()
```
case while accounting for graphs which we may not run backward of.


cc soumith voznesenskym penguinwu anijain2305 EikanWang jgong5 Guobing-Chen XiaobingSuper zhuhaozhe blzheng Xia-Weiwen wenzhe-nrv jiayisunx peterbell10 desertfire

[ghstack-poisoned]

ezyang commented on torch/_inductor/compile_fx.py:
```diff
@@ -215,6 +225,13 @@ def compile_fx_inner(
+                and not is_inference
+                and not is_backward
+            ):
+                boxed_forward_device_index.set(next(iter(graph.device_idxs)))
```
Why is it correct to only take the first device index?

eellison committed:
Update on "Account for forwards which whose corresponding backwards are not invoked"


Previously, when we would run a forward graph whose backward we never invoked it would prevent us from switching from warmup to recording. Now, refine the heuristic to allow incrementing the generation as soon as we invoke a backward graph. This still handles the 
```
mod1 = torch.compile(...)

mod2 = torch.compile(...) 

mod2(mod1(x)).sum().backward()
```
case while accounting for graphs which we may not run backward of. 

It also now handles the case where we skip cudagraphify the backward of a forward.


cc soumith voznesenskym penguinwu anijain2305 EikanWang jgong5 Guobing-Chen XiaobingSuper zhuhaozhe blzheng Xia-Weiwen wenzhe-nrv jiayisunx peterbell10 desertfire

[ghstack-poisoned]

eellison committed:
Update on "Account for forwards which whose corresponding backwards are not invoked"


Previously, when we would run a forward graph whose backward we never invoked it would prevent us from switching from warmup to recording. Now, refine the heuristic to allow incrementing the generation as soon as we invoke a backward graph. This still handles the 
```
mod1 = torch.compile(...)

mod2 = torch.compile(...) 

mod2(mod1(x)).sum().backward()
```
case while accounting for graphs which we may not run backward of. 

It also now handles the case where we skip cudagraphify the backward of a forward.


cc soumith voznesenskym penguinwu anijain2305 EikanWang jgong5 Guobing-Chen XiaobingSuper zhuhaozhe blzheng Xia-Weiwen wenzhe-nrv jiayisunx peterbell10 desertfire

[ghstack-poisoned]

eellison commented:
@pytorchbot merge

eellison commented:
@pytorchbot merge
